server/src/main/java/io/deephaven/server/table/ops/SnapshotTableGrpcImpl.java

```diff
@@ -3,76 +3,102 @@
+                throw GrpcUtil.statusRuntimeException(Code.UNIMPLEMENTED, String.format(
+                        "%s with %s (%d) == true does not support %s (%d), %s (%d), nor non-empty %s (%d)",
+                        descriptor.getFullName(),
+                        descriptor.findFieldByNumber(SnapshotWhenTableRequest.HISTORY_FIELD_NUMBER).getName(),
+                        SnapshotWhenTableRequest.HISTORY_FIELD_NUMBER,
+                        descriptor.findFieldByNumber(SnapshotWhenTableRequest.INITIAL_FIELD_NUMBER).getName(),
+                        SnapshotWhenTableRequest.INITIAL_FIELD_NUMBER,
+                        descriptor.findFieldByNumber(SnapshotWhenTableRequest.INCREMENTAL_FIELD_NUMBER).getName(),
+                        SnapshotWhenTableRequest.INCREMENTAL_FIELD_NUMBER,
+                        descriptor.findFieldByNumber(SnapshotWhenTableRequest.STAMP_COLUMNS_FIELD_NUMBER).getName(),
+                        SnapshotWhenTableRequest.STAMP_COLUMNS_FIELD_NUMBER));
+            }
         }
+    }
 
-        final String[] stampColumns = request.getStampColumnsList().toArray(CollectionUtil.ZERO_LENGTH_STRING_ARRAY);
-        final SelectColumn[] stampExpressions = SelectColumnFactory.getExpressions(request.getStampColumnsList());
-        ColumnExpressionValidator.validateColumnExpressions(stampExpressions, stampColumns, lhs);
-
-        final FunctionalInterfaces.ThrowingSupplier<Table, RuntimeException> doSnapshot =
-                () -> lhs.snapshot(rhs, request.getDoInitialSnapshot(), stampColumns);
-
-        final Table result;
-        if (!lhs.isRefreshing() && !rhs.isRefreshing()) {
-            result = doSnapshot.get();
-        } else {
-            result = updateGraphProcessor.sharedLock().computeLocked(doSnapshot);
+    @Override
+    public Table create(
+            final SnapshotWhenTableRequest request,
+            final List<SessionState.ExportObject<Table>> sourceTables) {
+        Assert.eq(sourceTables.size(), "sourceTables.size()", 2);
+        final Table base = sourceTables.get(0).get();
+        final Table trigger = sourceTables.get(1).get();
+        final Builder builder = SnapshotWhenOptions.builder();
+        if (request.getInitial()) {
+            builder.addFlags(Feature.INITIAL);
+        }
+        if (request.getIncremental()) {
+            builder.addFlags(Feature.INCREMENTAL);
+        }
+        if (request.getHistory()) {
+            builder.addFlags(Feature.HISTORY);
+        }
+        for (String stampColumn : request.getStampColumnsList()) {
+            builder.addStampColumns(ColumnName.of(stampColumn));
         }
-        return result;
+        try (final SafeCloseable _lock = lock(base, trigger)) {
+            return base.snapshotWhen(trigger, builder.build());
+        }
+    }
+
+    private SafeCloseable lock(Table base, Table trigger) {
+        return base.isRefreshing() || trigger.isRefreshing() ? updateGraphProcessor.sharedLock().lockCloseable() : null;
```

Comment:
To answer your verbal question, I think it would be best to prevent the users from ever using a static trigger table.

Reply:
I worry about breaking unit tests (specifically, python I think) and delaying merge of this. I'll do a quick pass to see if it's less disruptive than I think it might be. Otherwise, will follow-up with another ticket.

Reply:
Good news - only the python tests are breaking if we enforce trigger must be refreshing; which I was going to hopefully re-work w/ Jianfeng anyways.